### DISCUSSION TITLE
classes/class_multiplayerpeer

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#290] [User-contributed notes]
# classes/class_multiplayerpeer

Inherits: PacketPeer< RefCounted< Object Inherited By: ENetMultiplayerPeer, MultiplayerPeerExtension, OfflineMultiplayerPeer, WebRTCMultiplayerPeer, WebSocketMultiplayerPeer Abstract class for spec...

https://docs.godotengine.org/en/stable/classes/class_multiplayerpeer.html

**Comment by Magodra** [comment]
Should documentation state what happen with packages received out of order for TRANSFER_MODE_UNRELIABLE_ORDERED?
According to enet doc "For unreliable packets, ENet will simply discard the lower sequence number packet if a packet with a higher sequence number has already been delivered. This allows the packets to be dispatched immediately as they arrive, and reduce latency of unreliable packets to an absolute minimum. For reliable packets, if a higher sequence number packet arrives, but the preceding packets in the sequence have not yet arrived, ENet will stall delivery of the higher sequence number packets until its predecessors have arrived."

**Reply by Calinou** [reply]
MultiplayerPeer can be used with backends other than ENet (e.g. WebSockets, WebRTC), so I would consider this an implementation detail. Therefore, I think it's best not to mention it here.

**Comment by MatthewHinds** [comment]
It mentions

>"Note: The MultiplayerAPI protocol is an implementation detail and isn't meant to be used by non-Godot servers. It may change without notice.".

How would we then communicate between a Godot Client and a seperate C# console app using ENet-CSharp?

**Reply by Calinou** [reply]
Godot's high-level multiplayer API isn't meant to interoperate with other programs. It's designed around the concept of Godot nodes and its scene tree, so it wouldn't make as much sense to use it if the server side wasn't running on Godot. If you want interoperability, you should use another protocol depending on your needs (from highest-level to lowest-level: HTTP, WebSockets, WebRTC, raw ENet, raw TCP, raw UDP).

If you have a multiplayer server that runs using Godot but you want to interact with it from other programs (e.g. for remote administration), you can make it listen on several ports at the same time. One port is used for the actual game server functionality, while the other is used for an HTTP/WebSocket server or similar. Note that Godot doesn't come with HTTP server functionality out of the box, but you can use an add-on like https://github.com/Mr-Byte-Games/gd-http-server.

**Reply by TyGrze** [reply]
I personally use my own protocol and use Google's Protobuf to generate the message types and serialization code in any language I need

There are Protobuf compilers for GDScirpt and C#

https://github.com/oniksan/godobuf
https://github.com/protobuf-net/protobuf-net
